python/perf-kernels/fused_moe/moe-gemm.py

```diff
@@ -0,0 +1,495 @@
+        {"M": 4096, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+    ]
+    return configs
+
+
+def model_benchmark_configs(args):
+    config_file = args.model_configs
+    configs = get_model_configs(config_path=config_file, model_families=["mistral"], model=args.model)
+    moe_configs = []
+    M = args.M if args.M else 4096  # check size
+    # M, K, N, E, top_k
+
+    for model_name, config in configs.items():
+        N1 = config["intermediate_size"]
+        K1 = config["hidden_size"]
+
+        N2 = config["hidden_size"]
+        K2 = config["intermediate_size"] // 2
+
+        E = 8
+        top_k = 2
+        moe_configs.append((model_name, M, N1, K1, E, top_k))
+        moe_configs.append((model_name, M, N2, K2, E, top_k))
+
+    return moe_configs
+
+
+def run_benchmark(custom, args):
+    routed_weight = args.routed_weight
+    dtype = arg_to_torch_dtype[args.dtype]
+    x_names = ['M', 'N', 'K', 'E', 'top_k']
+    if custom:
+        assert args.M and args.N and args.K and args.E and args.top_k, \
+            "Please provide M, N, K, E, top_k for custom runs."
+        x_vals_list = [(args.M, args.N, args.K, args.E, args.top_k)]
+    else:
+        if args.model:
+            x_vals_list = model_benchmark_configs(args)
+            x_names = ['model', 'M', 'N', 'K', 'E', 'top_k']
+        else:
+            configs = get_configs()
+            x_vals_list = [(cfg['M'], cfg['N'], cfg['K'], cfg['E'], cfg['top_k']) for cfg in configs]
+
+    line_names = ['Time (ms)', 'TFLOPS', 'Bandwidth (GB/s)']
+    line_vals = ['time', 'tflops', 'bandwidth']
+
+    benchmark = triton.testing.Benchmark(x_names=x_names, x_vals=x_vals_list, line_arg='metric', line_vals=line_vals,
+                                         line_names=line_names, styles=[('red', '-'), ('blue', '-'), ('yellow', '-')],
+                                         ylabel='ms / TFLOPS / GB/s', plot_name='moe-gemm-benchmark',
+                                         args={'dtype': dtype, 'routed_weight': routed_weight})
+
+    @triton.testing.perf_report([benchmark])
+    def bench_moe_gemm(M, N, K, E, top_k, dtype, routed_weight, metric, model=None):
+        # metric will be either 'time'/'tflops' or 'bandwidth'
+        a, b, c, topk_weights, topk_ids, sorted_token_ids, expert_ids, num_tokens_post_padded, config = input_helper(
+            M, N, K, top_k, E, routed_weight=routed_weight, dtype=dtype)
+
+        # (M, K) * (top_k, N, K) -> (M, top_k, N). 2 for multiplication and accumulation
+        flops = 2.0 * M * top_k * K * N
+        # The weight is applied on the gemm product which has the shape of (M, top_k, N)
+        if routed_weight:
+            flops += M * top_k * N
+
+        bytes_ = torch.tensor([], dtype=dtype).element_size()
+        # (M, K) memory load for A (E,  N,  K) for B not (top_k,  N,  K) because we are in total bringing in all expert matrices into the chip from memory. It's just that not all multiply the same A.
```

Comment:
Is it possible that only n (n < 8) experts are chosen for all of the M tokens?
If so, is the kernel still loading the other unused 8-n experts?

Reply:
It might happen when M is very small. But in most cases, especially when benchmarking with a reasonable batch size, each the expert is chosen by some token, which is where the "sparsity" idea is from

Reply:
So when M is very small, the kernel can have higher performance in terms of GB/s than it's real performance?
Did you check if this is done in the vllm for both AMD and NV backend?

Reply:
I don't think they have done such calculation in https://github.com/ROCm/vllm/blob/develop/benchmarks/kernels/benchmark_moe.py. I had it there for tuning purpose. But yes, when M is small enough, it might miscalculate the wrong memory usage. Although in the real use cases you won't have that small M most likely